### PR TITLE
gui: Fix boost::signals2::no_slots_error in early calls to InitWarning

### PR DESCRIPTION
**Pull request by promag**
Adding the following to `bitcoin.conf`
```
[xxx]
disablewallet=1
```
And running `bitcoin-qt` gives:
```
libc++abi.dylib: terminating with uncaught exception of type boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::signals2::no_slots_error> >: boost::signals2::no_slots_error
```

Fixes regression in #14708.

### CODE REVIEW

File: src/qt/bitcoin.cpp [hunk 1]
```diff
@@ -691,6 +691,10 @@ int main(int argc, char *argv[])
 #endif
     // Install qDebug() message handler to route to debug.log
     qInstallMessageHandler(DebugMessageHandler);
+
+    // Redundantly log and print message in non-gui fashion
+    auto handler_message_box = node->handleMessageBox(noui_ThreadSafeMessageBox);
```
[the remark this answers is not in the document]
Comment (promag) [reply src/qt/bitcoin.cpp]: Internally `handler_message_box` has a `boost::signals2::scoped_connection` and the destructor disconnects.

[the remark this answers is not in the document]
Comment (laanwj) [reply src/qt/bitcoin.cpp]: might want to mention this in the comment, that is quite the gotcha

**Comment by promag** [comment]
Alternatively could call `noui_connect` in `src/qt/bitcoin.cpp`?

**Comment by laanwj** [comment]
What will happen with the return value if you register multiple handlers?

**Comment by laanwj** [comment]
I think the safest solution to this is to not call the message box at all for unrecognized sections. Simply call `LogPrintf` directly. That's what will happen after all!

**Comment by promag** [comment]
Updated, @MarcoFalke you might want to check this out since you reviewed #14708.

**Comment by laanwj** [comment]
utACK 3a6a48929d49186598b2e8f16936de903c1f1ba1

So to be clear, the problem is that at this stage of initialization, we're not yet able to show message boxes in Qt [1]. @promag's initial fix changed the slot handler to first register a message box handler that only logs, but, it seems safer to simply change these `InitWarning` calls to `LogPrintf` as that's what is going to happen both for the UI and `bitcoind`.

1. The reason is that there are some things concerning locale and language setup that only can run *after* configuration parsing, after which the main UI window is created. It might be possible to refactor this at some point but for fixing the issue, this is the safer option.

**Review by maflcko — DISMISSED** [review]
Slightly tend to NACK. This makes the code fragile in light of refactoring such as move-only commits, since it is not clear when `InitWarning` is safe to call. Wouldn't it be easier to always register the no_ui listener to the signal as soon as possible in the initialization sequence (regardless of `bitcoind` vs `bitcoin-qt`) and then optionally, if the gui is run, also register the gui message box listener?

Side-note: Previously we wouldn't even log nor print warnings when running the gui, but that is now fixed (see #14162)

**Comment by promag** [comment]
Previous fix in 91ce5a2, which @MarcoFalke disapproved.

Current fix does what was suggested in https://github.com/bitcoin/bitcoin/pull/14783#issuecomment-441016502.

**Comment by promag** [comment]
Not sure about this because I think it should go thru `interfaces::Node`?

**Review by maflcko — COMMENTED** [review]
Should remove the now-redundant (?) `static void InitMessage` in ` src/qt/bitcoin.cpp`?

**Comment by laanwj** [comment]
I'm still not sure about this, but ok…

**Comment by promag** [comment]
@laanwj I'm not ignoring them, I'm trying to address them.

**Comment by promag** [comment]
@laanwj you mean noui_register instead of noui_register`. Regarding:

> * Does noui_register register anything else that might interfere with the GUI?

`noui_connect` only connects 3 slots that only call `LogPrintf` and `fprintf`, so I think we are safe.

> * What about the return argument of `ThreadSafeMessageBox`, will registering multiple handlers cause an issue here (similar for _other_ handlers that `noui_register` registers)

In boost::signals, slot results are combined to give the signal result. The default combiner is to give the last slot result. So the first connected slots are those in `noui_connect` and later are the UI slots, which means the signal result comes from the UI slots. So it's safe to say there is no behavior change.

File: src/qt/bitcoin.cpp [hunk 2]
```diff
@@ -563,6 +558,11 @@ int main(int argc, char *argv[])
 
     std::unique_ptr<interfaces::Node> node = interfaces::MakeNode();
 
+    // Subscribe to global signals from core
+    std::unique_ptr<interfaces::Handler> handler = node->handleMessageBox(noui_ThreadSafeMessageBox);
+    std::unique_ptr<interfaces::Handler> handler = node->handleQuestion(noui_ThreadSafeQuestion);
+    std::unique_ptr<interfaces::Handler> handler = node->handleInitMessage(noui_InitMessage);
```
Comment (MarcoFalke) [inline src/qt/bitcoin.cpp]: This does not compile.

**Comment by jonasschnelli** [comment]
Concept ACK

File: src/qt/bitcoin.cpp (hunk 2 above)
Comment (promag) [reply src/qt/bitcoin.cpp]: Fixed.

**Comment by promag** [comment]
> Should remove the now-redundant (?) static void InitMessage in src/qt/bitcoin.cpp?

@MarcoFalke done.

File: src/qt/bitcoin.cpp [hunk 3]
```diff
@@ -563,6 +558,11 @@ int main(int argc, char *argv[])
 
     std::unique_ptr<interfaces::Node> node = interfaces::MakeNode();
 
+    // Subscribe to global signals from core
+    std::unique_ptr<interfaces::Handler> handler_message_box = node->handleMessageBox(noui_ThreadSafeMessageBox);
+    std::unique_ptr<interfaces::Handler> handler_question = node->handleQuestion(noui_ThreadSafeQuestion);
+    std::unique_ptr<interfaces::Handler> handler_init_message = node->handleInitMessage(noui_InitMessage);
+
     // Do not refer to data directory yet, this can be overridden by Intro::pickDataDirectory
```
Comment (MarcoFalke) [inline src/qt/bitcoin.cpp]: What makes it safe to subscribe to them here already, when some of then log to the debug log, which is in the datadir?

Comment (promag) [reply src/qt/bitcoin.cpp]: If that was the case then `boost::signals2::no_slots_error` would have been raised before?

**Comment by MarcoFalke** [comment]
utACK 6bbdb2077eb4dae9f7a1938bdbcffab884719cbe. Going to test the next time I compile the gui

**Comment by Sjors** [comment]
I can reproduce the error on macOS 10.14.1 and confirm 6bbdb20 fixes it. It shows a `Section [xxx] is not recognized.` after the command, but someone launching QT via a desktop icon wouldn't see that.

Would be nice to have to test, though that might require #11625

**Comment by promag** [comment]
@Sjors thanks, and the improvement can come next.

**Pull request merged**